File: schemas/multi-language-i18n/drizzle/translation_memory_entries.ts
// translation_memory_entries: Reusable translation pairs for translation memory.
// See README.md for full design rationale.

import { pgTable, uuid, text, integer, numeric, timestamp, index } from "drizzle-orm/pg-core";
import { locales } from "./locales";
import { users } from "../../auth-rbac/drizzle/users";

export const translationMemoryEntries = pgTable(
  "translation_memory_entries",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    sourceLocaleId: uuid("source_locale_id")
      .notNull()
      .references(() => locales.id, { onDelete: "cascade" }),
    targetLocaleId: uuid("target_locale_id")
      .notNull()
      .references(() => locales.id, { onDelete: "cascade" }),
    sourceText: text("source_text").notNull(),
    targetText: text("target_text").notNull(),
    sourceHash: text("source_hash").notNull(),
    domain: text("domain"),
    qualityScore: numeric("quality_score", { precision: 3, scale: 2 }),
    usageCount: integer("usage_count").notNull().default(0),
    source: text("source").notNull().default("human"),
    createdBy: uuid("created_by").references(() => users.id, {
      onDelete: "set null",
    }),
    createdAt: timestamp("created_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
    updatedAt: timestamp("updated_at", { withTimezone: true })
      .notNull()
      .defaultNow()
      .$onUpdateFn(() => new Date()),
  },
  (table) => [
    index("idx_translation_memory_entries_source_locale_id_target_locale_id_source_hash").on(
      table.sourceLocaleId,
      table.targetLocaleId,
      table.sourceHash
    ),
    index("idx_translation_memory_entries_source_locale_id_target_locale_id").on(
      table.sourceLocaleId,
      table.targetLocaleId
    ),
    index("idx_translation_memory_entries_domain").on(table.domain),
  ]
);

File: schemas/multi-language-i18n/spacetimedb/translation_memory_entries.rs
// translation_memory_entries: Reusable translation pairs for translation memory.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

// Composite index: (source_locale_id, target_locale_id, source_hash) — enforce in reducer logic

#[spacetimedb::table(name = translation_memory_entries, public)]
pub struct TranslationMemoryEntry {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub source_locale_id: String, // FK → locales.id (cascade delete)

    #[index(btree)]
    pub target_locale_id: String, // FK → locales.id (cascade delete)

    pub source_text: String,
    pub target_text: String,
    pub source_hash: String,

    #[index(btree)]
    pub domain: Option<String>,

    pub quality_score: Option<f64>,
    pub usage_count: i32,
    pub source: String,
    pub created_by: Option<String>, // FK → users.id (set null)
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
